### ISSUE TITLE
Terrain tool "glitch"

### ISSUE THREAD

**Issue by coNNecTT** [bjorn/tiled#1172]
Hello, I'm trying to make the terrain tool work for a while now, but I keep getting into this problem. I couldn't find any information about this problem. I'm trying to use 3 and more terrain types on one layer. It works, but when drawing on an edge of a terrain, weird things start to happen. Here are 2 screenshots of my test map with my cursor on the same tile (the "glitch" is random every time i hover over the tile): 
http://puu.sh/mp5H5/3bd8505b2d.jpg
http://puu.sh/mp5Je/3e5b6d27a7.jpg
I think this has something to do with probability on the tiles (every tile is set to 1 though). I think it would help out many people if the terrain tool had some examples (on how to setup the tileset for the terrain tool of course) and a more in-depth explanation of how it works. I saw a video where the terrain tool does what I'm trying to do without any problems. I can show some screenshots of my test tileset's terrain information if needed.

So basically, what exactly causes this "glitch" and how could it be prevented? Is it a bug? (I don't think so)

**Comment by bjorn** [comment]
This glitch happens for two reasons:

* Your tileset is probably missing certain combinations that are possible when combining three terrains.
* The Terrain Tool algorithm only looks at each cell once. So when it has picked a certain fitting tile, it will not go back and reconsider it when it gets stuck further down the path.

While I would like to improve the tool eventually to have it make smaller changes, this can easily get very computationally intensive and memory hungry. So really the best solution for now is to make sure you got all possible combinations in your tileset.

Here's an example of a tileset that avoids having partial sets of transitions:

https://github.com/tales/sourceoftales/blob/master/tiles/terrain.png

**Comment by coNNecTT** [comment]
Thanks! Indeed I was missing quite a lot of tiles and I somehow never realised it. I got it working perfectly and I want to share a template file that I made for all the transitions.
https://www.dropbox.com/s/p3ss0770fkke2l7/TileTransitionTemplate.rar?dl=1

**Comment by bjorn** [comment]
That definitely looks useful. I've uploaded one version here as plain image:

![tiletransitiontemplate16px](https://cloud.githubusercontent.com/assets/531764/12217815/811d85ea-b70c-11e5-8153-0c072f88d578.png)

Note also that since Tiled 0.15, there is a Terrain Generator tool that helps to make all the combinations out of a set of transitions from each terrain type to "transparency". I'll aim to document this tool soon, or at least before reaching Tiled 1.0.

**Comment by phil123456** [comment]
please dont explain where is the terrain generator tool, that could be usefull

**Comment by bjorn** [comment]
@phil123456 You can find its source code at https://github.com/bjorn/tiled/tree/master/src/terraingenerator and it generally ships with Tiled as `terraingenerator.exe`. I should really put some time aside to document this tool soon...
